File: python/src/basket_informations.py
class BasketInformations:
    # The product of the basket
    map = {}

    # The fact that the basket has promo code
    code_de_promotion = False

    def add_product_to_basket(self, product, price, is_promo_code):
        if is_promo_code:
            BasketInformations.code_de_promotion = True
        else:
            BasketInformations.map[product] = price

    def get_basket_price(self, in_cents):
        v = 0
        for s in BasketInformations.map.values():
            v += s
        if BasketInformations.code_de_promotion:
            v -= 100
        return v * 100 if in_cents else Long(v)

    def reset_basket(self):
        self.buy_basket()
        BasketInformations.code_de_promotion = False

    def buy_basket(self):
        BasketInformations.map.clear()

    def is_basket_contains(self, produit):
        found = False
        for s in BasketInformations.map.keys():
            if s == produit:
                found = True
                break
        return found

    def mix_with_basket(self, b):
        BasketInformations.map.update(b.map)

File: scala/src/BasketInformations.scala
import scala.collection.mutable.HashMap

class BasketInformations {
  // The product of the basket
  val map: HashMap[String, Int] = HashMap()

  // The fact that the basket has promo code
  private var codeDePromotion: Boolean = false

  def addProductToBasket(product: String, price: Int, isPromoCode: Boolean): Unit = {
    if (isPromoCode) {
      codeDePromotion = true
    } else {
      map += (product -> price)
    }
  }

  def getBasketPrice(inCents: Boolean): Long = {
    var v: Int = 0
    for (s <- map.values) {
      v += s
    }
    if (codeDePromotion) {
      v -= 100
    }
    if (inCents) {
      v * 100
    } else {
      v.toLong
    }
  }

  def resetBasket(): Unit = {
    buyBasket()
    codeDePromotion = false
  }

  def buyBasket(): Unit = {
    map.clear()
  }

  def isBasketContains(produit: String): Boolean = {
    var found = false
    for (s <- map.keys) {
      if (s == produit) {
        found = true
      }
    }
    found
  }

  def mixWithBasket(b: BasketInformations): Unit = {
    map ++= b.map
  }
}
